### ISSUE TITLE
Nvidia CP frame limiter leads to high MsInPresentAPI values 

### ISSUE THREAD

**Issue by DevTechProfile** [GameTechDev/PresentMon#73]
[A user on reddit](https://www.reddit.com/r/allbenchmarks/comments/f5l6pk/nvidias_control_panel_fps_limiter_vs_rivatuner_vs/) compared different frame limiters, including NV CP and RTSS. CapFrameX provides input lag approximation based on your formular given in the GitHub description. 

```LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)```

The problem is that the values for MsInPresentAPI are very high, so the overall latency becomes very low. This is only the case if the frame limiter from Nvidia is used. What's going wrong here? Is there a problem at all?

Capture data with NV CP frame limiter: https://pastebin.com/0ipbcaKs
Capture data with RTSS frame limiter: https://pastebin.com/ueSpNSWt

**Comment by RodroG-85** [comment]
Hello. I am the author of the analysis referenced by @DevTechProfile if you need all the record files I can send you.

Just as note: Only the WolfYB (Vulkan) results seems correct to me in terms of approximate input lag.

Thanks in adavance.

**Comment by JeffersonMontgomery-Intel** [comment]
The difference is based on where the delay is added.
- If a delay is added by the OS/driver while Present() is running, then the delay will be included in the MsInPresentAPI metric
- If Present() is hooked, or the application does this itself, and the delay added before Present() is called then it will be included in the MsBetweenPresents metric.
In both of these cases, the latency equation above will be correct (assuming the workload samples input and simulation time when Present() returns).

If the delay is added after Present() returns, but before the application samples input/time, then the equation will be wrong (too large) since that assumption is then incorrect.

**Comment by DevTechProfile** [comment]
So, in the case of the NV limiter the equation could be wrong and there is no way to validate this reliably? But the equation is too small (low latency) not too large. Do you think that the NV limiter can cause such behaviour?

**Comment by RodroG-85** [comment]
That's the question and I'd say the answer too.

The use of a different FPS limiter was the independent variable in each
test, so such wrong results for the equation, would only be explained by
the use of the NV CP limiter.

**Comment by JeffersonMontgomery-Intel** [comment]
No, the NVIDIA numbers are correct.

I don't know how RivaTuner or these applications implement their limiters, but for them there is two possibilities.  Either the limiters are actually adding latency (delaying before Present()) or the calculation above is wrong (if they delay after Present() and before the next frame's input).

For RivaTuner, we could test which is the case writing a simple app that just Present()'s and times itself so we'd see where the delay is being introduced.  The delay might even be visible in WPA/gpuview which could be a more general way to determining which case it is.

**Comment by RodroG-85** [comment]
Recently, I read from Unwinder, the dev of RTSS, that the RTSS limiter hook
would work, at some point, similar to this sequence(will report here later
his exact words on this): "present and THEN sleep"...

**Comment by JeffersonMontgomery-Intel** [comment]
OK, then the above equation will be wrong.  You'd want:

`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI) - how_ever_long_rtss_slept`

I'm not sure how to figure out that last part at the moment.  It's possible RTSS computes it based on MsBetweenPresents, and you could try to reproduce that.  Or like I mentioned above, try to detect the sleep in the ETW trace to subtract it out.

Ideally PresentMon could get more-direct visibility into when apps sample input, but for now it relies on that assumption above which turns out to be wrong in this case unfortunately.

**Comment by RodroG-85** [comment]
I think the best would be to ask him and involve Unwinder in this issue.

**Comment by RodroG-85** [comment]
From Unwinder, recently, in [guru3D forum](https://forums.guru3d.com/threads/nvidia%E2%80%99s-new-control-panel-fps-limiter-vs-rivatuner-vs-in-engine.430198/page-4#post-5751598):

> It doesn't work that way. It doesn't sleep until present. It presents the frame immediately THEN sleeps. 

And [this](https://forums.guru3d.com/threads/nvidia%E2%80%99s-new-control-panel-fps-limiter-vs-rivatuner-vs-in-engine.430198/page-5#post-5751745):

> I don’t need GPUView to reverse engineer my own implementation. I've replied to two of your similar posts in that youtube video comments and here and I'll repeat the same again: you're misunderstanding the way FPS limtier is implemented. Busy wait is NOT performed between application's rendering and present calls, busy wait loop is performed AFTER calling original present() from hook. Please let's avoid turning this thread into the next false RTSS understanding discussion. You're seeing it wrong, sorry.
> 
> > If that worked that way then you should have no input lag since the cmd to screen would have 
> no additional latency
> 
> Present() calls are asynchronous and queued by driver and DX runtimes. It always have some latency and you NEVER have the frame presented on screen immediately after present call. 

An also [this](https://forums.guru3d.com/threads/nvidia%E2%80%99s-new-control-panel-fps-limiter-vs-rivatuner-vs-in-engine.430198/page-5#post-5751873):

> Once again, it DOESN'T work that way. Any third party framerate limiter (including the driver's internal one), which is limiting the framerate OUTSIDE the game engine, is always doing the limiting INSIDE application's 3D API present() call. It will always look like LONG and delayed present() call (ending exactly before the next frame is about to be presented) in tools tracing and visualizing application's 3D API call timings. But you do not and cannot visualize what is going on inside the hook. And it is always relaying control to real present() THEN waiting for proper timestamp to return control to application's renderer thread. Let's stop on that PLEASE, I spent all past weekend day and night on commenting RTSS implementation related misunderstanding and I'd really like to avoid doing the same again. 

An [this](https://forums.guru3d.com/threads/nvidia%E2%80%99s-new-control-panel-fps-limiter-vs-rivatuner-vs-in-engine.430198/page-5#post-5753177) from other user replaying to a confused user on the above:

It looks something like this (pseudocode):
present_hook() // hook
```
{
    present(); // real
    wait(timestamp);
    return;
}
```
You're suggesting it looks something like this:
present_hook() // hook
```
{
    wait(timestamp);
    present(); // real
    return;
}
```
But you won't be able to tell using GPUView since the hook present() call is taking the same amount of time.

However, the implementation is different. Unwinder is using the first. @Unwinder, correct me if I'm wrong.

And [Unwinder](https://forums.guru3d.com/threads/nvidia%E2%80%99s-new-control-panel-fps-limiter-vs-rivatuner-vs-in-engine.430198/page-5#post-5753199) replied:

> Correct.

**Comment by JeffersonMontgomery-Intel** [comment]
I can't tell if you're trying to suggest something other than: here's a bunch more quotes of unwinder saying RTSS sleeps after the present call?

**Comment by RodroG-85** [comment]
> I can't tell if you're trying to suggest something other than: here's a bunch more quotes of unwinder saying RTSS sleeps after the present call?

Sorry, what? Just quoting his exact word as I said above: the comments where he explained how his limiter works. I added an user comment because I think it can help to represent his implementation. Just this. Consecuences of this on your equation, would be on your side though.

**Comment by RodroG-85** [comment]
Hello. First of all, I would like to thank @JeffersonMontgomery-Intel  and @DevTechProfile  for the interest you shown in this issue raised from the approximate input lag results of my analysis.

However, reading again all the above I still have almost same doubts or even more on the issue. I suppose it has to do with the fact that I am not a developer and I do not have enough prior and technical knowledge to properly represent the issue and deduce the appropriate practical consequences in a practical and real analysis contexts. Therefore, I would highly appreciate if you could clarify and confirm me if, in the particular case of my analysis, the equation was correct or wrong (in that case what changes in the equation woulbe required) in each of the FPS limitation scenarios (I mean: **in-engine**, **RTSS limite**r, **NV CP limiter**) .

It depends on this that I continue using this input lag approximation approach in my benchmarkings and that I can consider it as an alternative, approximate but reliable and valid, method when estimating or valuing latency in gaming contexts such as the one was used (G-Sync On + NV V-Sync ON + [FPS Limiter]) or others.

Thank you in advance and, please, excuse my current limited understanding in this regard.

**Comment by DevTechProfile** [comment]
As a consequence I will add a note to the Input Lag Approximation in CapFrameX that the method does not yield correct values when using the RTSS limiter. This does not mean that RTSS is not working correctly, but in the equation another summand is missing to map the offset latency (order present() and wait(timestamp)). I hope that Jefferson can implement some kind of detection.

**Comment by RodroG-85** [comment]
> As a consequence I will add a note to the Input Lag Approximation in CapFrameX that the method does not yield correct values when using the RTSS limiter. This does not mean that RTSS is not working correctly, but in the equation another summand is missing to map the offset latency (order present() and wait(timestamp)). I hope that Jefferson can implement some kind of detection.

Thank you @DevTechProfile So, are you sure the current method yeild correct values to In-Engine and NV CP limiters too? I am asking becuase the following Unwinder words make me think that it would only do so in the case of the In-Engine limiter (or maybe I am wrong and I have not interpreted his words correctly):

> Any third party framerate limiter (including the driver's internal one), which is limiting the framerate OUTSIDE the game engine, is always doing the limiting INSIDE application's 3D API present() call. It will always look like LONG and delayed present() call (ending exactly before the next frame is about to be presented) in tools tracing and visualizing application's 3D API call timings. But you do not and cannot visualize what is going on inside the hook. And it is always relaying control to real present() THEN waiting for proper timestamp to return control to application's renderer thread.

**Comment by DevTechProfile** [comment]
You want in-game and RTSS to produce similar results, right? Then the equation only fits the approach of Nvidia. Oh man...

**Comment by RodroG-85** [comment]
> You want in-game and RTSS to produce similar results, right? Then the equation only fits the approach of Nvidia. Oh man...

I really don't want anything mate, just to be sure for which FPS limiter scenarios the @JeffersonMontgomery-Intel 's current equation is correct of wrong and how this can affect its use in my analysis.

Unwinder seems to be saying that "any third party framerate limiter (including the driver's internal one)", that is, the NV CP limiter in this context, would be working in the same way as his RTSS FPS limiter is doing. Therfore, for me, isn't yet clear if the current equation could be wrong for both the RTSS and NV CP limiters.

**Comment by DevTechProfile** [comment]
But the RTSS vs. NV CP MsInPresentAPI numbers are completely different!

**Comment by RodroG-85** [comment]
> But the RTSS vs. NV CP MsInPresentAPI numbers are completely different!

Yes, I know. And In-Engine vs RTSS were more close... Please, I need a clear and based answer for each FPS limiter condition. Something like this:

_In-Engine FPS limiter (approximate input lag equation):_
`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)` (if correct according with results)

or

`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI) - how_ever_long_in-engine_slept ` (if wrong according to results)

_RTSS FPS limiter (approximate input lag equation):_
`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)` (if correct according with results)

or

`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI) - how_ever_long_rtss_slept ` (if wrong according to results)

_NV CP FPS limiter (approximate input lag equation):_
`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)` (if correct according with results)

or

`LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI) - how_ever_long_nvcp_slept ` (if wrong according to results)

**Comment by jenatali** [comment]
At the end of the day, latency should be measured from the time the engine sampled user input, to the time that the result of that input is visible to the user. This is not possible to measure using software, without knowing the duration between sending pixels to the display, and the light reaching the user. Further, there exists no *standard* way to detect when the engine sampled input, given how many different kinds of input exist.

What PresentMon can measure is the amount of time from when the engine *finished* a frame on the CPU, to the point when the OS requested the display to start scanning that frame. If you add an assumption that the engine sampled input immediately after finishing the previous frame, you end up with the equation you started with.

Frame limiters *other than ones implemented inside the D3D runtime or driver* break this assumption, because they insert a delay between finishing one frame, and sampling input for the next. That causes the equation you have been using to be incorrect for them, unless you could also know how long they slept. This is something else that is not knowable in a *standard* way.

In summary:
* The NVCP latency should be (more or less) accurate according to that equation.
* RTSS is definitely not accurate according to that equation.
* For in-engine limiters, well, each engine can implement it themselves, so it depends whether they sleep:
  * After rendering and before presenting: your equation would be accurate, and show a high latency.
  * After presenting before rendering the next frame: your equation would be *innacurate*, and show a high latency.

**Comment by aggieNick02** [comment]
I've been reading and digesting this, and it seems like there is a little ambiguity in what PresentMon reports which would be helpful to clear up and then could let us treat this a little more formally (or at least help me follow it). Apologies for the text variables that follow.

Let's say we have two frames we're looking at, 1 and 2. The present for 1 starts at p_1s and ends at p_1e. The present for 2 starts at p_2s and ends at p_2e. Frame 1 displays at d_1 and frame 2 at d_2.

So the timeline looks like this:
---p_1s---p_1e---d_1---p_2s---p_2e---d_2

The original approximation is:
LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)

Assuming the sampling for frame 2 starts immediately after p_1e, the time between when sampling for frame 2 starts and frame 2 hits the screen is:

d_2 - p_1e

If the logged present times, between presents, and until displayed are all measured from when the present starts, then
MsBetweenPresents = p_2s - p_1s
MsUntilDisplayed = d_2-p_2s
previous(MsInPresentAPI) = p_1e - p_1s
@JeffersonMontgomery-Intel : Are these the correct meanings of the columns?

and

LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)
 = p_2s-p_1s + d_2-p_2s - (p_1e - p_1s)
 = d_2 - p_1e

as desired. 

To be able to deal with the frame limiters, we need to know when the times for the trace events are logged. I'm guessing they are logged by the OS before and after the call to present, so that if present is hooked, they are measuring the hooked function, not the original present. So the MsInPresentsAPI will increase dramatically/artificially.

**Comment by jenatali** [comment]
Your understanding is correct, except for where the event are logged. They are logged *after* any API hooks call into the OS binaries. In other words, call flow goes App -> Hook -> OS -> Driver for entering Present, and Driver -> OS -> Hook -> App when exiting, so the Present start/end events will include the limiter's wait only if the driver or OS is the layer that does the wait. If the hook does it, the wait won't be captured in the start/end events.

**Comment by aggieNick02** [comment]
Thanks @jenatali. I had it flipped, and your explanation makes sense.

**Comment by aggieNick02** [comment]
With the variables defined above, I want to take a quick look at the original assumption all of this latency approximation is based on. That is, that the latency is approximately 

d_2 - p_1e

and assumes sampling starts immediately when the present completes. I don't know anything about how real game engines do their sampling, but by the time we start a Present(), I'm pretty sure the input sampling for that frame has stopped. Which means the input for frame 2 starts, at the latest, at p_1s, not p_1e. Even if we don't look at the input until p_1e, there is input that occurred between p_1s and p_1e that still has to be processed. So the latency is at least

d_2 - p_1s = MsBetweenPresents + MsUntilDisplayed.

That is, I don't understand how/why we subtract the previous(MsInPresentAPI) when trying to approximate latency. Am I missing something? I want to be sure the original approximation equation is right before we examine what happens when a frame limiter is introduced.

**Comment by jenatali** [comment]
Assuming a single-threaded engine, the app can't do anything else while in Present, so it doesn't start doing frame 2 until frame 1 is done from the CPU side: that is, until p_1e. Even multithreaded engines generally fan out to render a frame, then collapse back to single-threaded at the end to Present it, rather than finishing frame 1 and starting frame 2 in parallel.

There's another variable which you haven't shown, which is when frame 1 is done from the GPU side, which is asynchronous and can happen while frame 2 is being recorded. Perhaps that's what you were thinking of?

**Comment by aggieNick02** [comment]
Sorry for the naive questions on this, and thanks for your patience. I'm certain I'm missing something in my understanding. I get that it doesn't start sampling input until p_1e, but there is still input from at least p_1s that hasn't been processed. Ignoring hooks/frame limiters, shouldn't the latency calculation try to approximate how much time elapses from when a user input occurs until it is reflected on screen, not from when that input is processed by the engine until it is reflected on screen? Put another way, why should 

To take this to an extreme, let's say we've got a single-threaded engine where every Present takes an entire second. MsBetweenPresents will be a little more than 1 second. MsUntilDisplayed will be a little more than 1 second. And MsInPresentAPI will be 1 second.

If I push a button right after the first present starts, it will not be in the first output frame, but in the second, which is a latency of about 2 seconds. But the original formula would give ~ 1 + 1 - 1 = 1 second for the latency.

**Comment by jenatali** [comment]
Oh sure, that's fair. I've been thinking of continuous input (like mouse movement, or a VR headset position) which gets sampled and accounts for all action since the last sample. But yes, for a discrete input like a button, the latency measurement may include up to an additional frame, depending on how the timing of the button press lines up with the period of input sampling.

**Comment by JeffersonMontgomery-Intel** [comment]
Yes, this is not the true physical button press-to-photons latency time.  This is the best possible approximation to it we can make with the metrics we have, making assumptions about the typical game architecture, and ignoring parts of the system (such as mouse/keyboard/monitor latencies) which can be important but we just have no insight into.  If you want accurate latency, you need to do use something like a high-speed camera watching the input & screen.

In terms of the input sample time assumption, I'd say sampling at the beginning of the frame is pretty typical.  You need to know the input and time in order to simulate the world and figure out what to draw etc. so it happens before most of the other frame work.

I think a lot of the other confusion is around how the Present() hooking works.  As Jesse said, PresentMon only instruments DXGI's Present() implementation which, in the RTSS case, is not the same as the "Present()" that the app calls.  In that case, the app is really calling RTSS's version of Present() which includes a delay that PresentMon doesn't see.

Here's a bit of a graphical representation (though, not showing GPU render/queueing/composition/etc. which is the same for all these cases):

```
App:        Present()--------------------------Input()--D3DCommands()---Present()
RTSS:        Present()-----Sleep()-------------
OS/Driver:    Present()---
PresentMon:   p_1s        p_1e
LatencyCalc:              |---------------------------------------------------------->

(in-game)
App:          Present()---Sleep()--------------Input()--D3DCommands()---Present()
OS/Driver:    Present()---
PresentMon:   p_1s        p_1e
LatencyCalc:              |---------------------------------------------------------->

App:        Present()--------------------------Input()--D3DCommands()---Present()
OS/Driver:  Present()---Sleep()----------------
PresentMon: p_1s                               p_1e
LatencyCalc:                                   |------------------------------------->
```

Note also that sleeping after Present() is the right implementation for a limiter, because you don't want the delay to _really_ add latency.  But, it causes the sleep to be included in the latency calculation leading to too large of a number.

**Comment by RodroG-85** [comment]
> In summary:
> 
> The NVCP latency should be (more or less) accurate according to that equation.
> RTSS is definitely not accurate according to that equation.
> For in-engine limiters, well, each engine can implement it themselves, so it depends whether they sleep:
> 
> After rendering and before presenting: your equation would be accurate, and show a high latency.
> After presenting before rendering the next frame: your equation would be innacurate, and show a high latency.

Thanks for this @jenatali . Therefore, from a practical point of view, in the context of my analysis, and without modifying the equation for the calculation of the approximate input lag, should I only consider the results relative to the NVCP limiter as valid?

**Comment by aggieNick02** [comment]
Thanks both @JeffersonMontgomery-Intel @jenatali for going through all of these details. I agree for current input position to output frame latency, this is the best that can be done without additional hardware. If you are looking for button press to output frame latency, without additional hardware, the best estimate is slightly different - you basically have to take the original latency and then add in the fact that the button press reflected in frame 2 could have happened any time after the sampling for frame 1 (at p_0e) until the sampling for frame 2 (at p_1e).

This means that button press latency, ignoring physical/electrical paths, could be as much as:
first_latency + p_1e - p_0e
= (d_2 - p_1e) + (p_1e - p_0e)
= d_2 - p_0e

For the numbers provided by PresentMon, this is:
d_2-p_0e
= d_2 - p_2s + p_2s - p_1s + p_1s - p_0s + p_0s - p_0e
= (d_2 - p_2s) + (p_2s - p_1s) + (p_1s - p_0s) + (p_0s-p_0e)
= MsUntilDisplay_frame-2 + MsBetweenPresents_frame-2-to-1 + MsBetweenPresents_frame-1-to-0 - MsInPresent_frame-0

**Comment by aggieNick02** [comment]
The more I think about this, the more I think just dropping the subtraction of the MsInPresent from the latency formula makes sense. I understand its purpose when we are talking current VR headset position to display latency. But it isn't applicable for button press type input.

Further, the true time in Present, when not incorrectly skewed by an OS/driver rate limiter, is generally an order of magnitude or two smaller than the other numbers involved in the formula - e.g. 16 ms vs 0.07ms.

I feel like the more appropriate thing to do is to provide an upper-bound on the latency caused by the software, which would be

MsUntilDisplay + MsBetweenPresents + prev (MsBetweenPresents)

and a lower bound, which (ignoring time in present) would be

MsUntilDisplay + MsBetweenPresents

Using these formulas would work with RTSS, hooking, NV CP, and non-limited (normal) gameplay.

**Comment by JeffersonMontgomery-Intel** [comment]
Yeah that's a reasonable upper bound.  MsInPresentAPI can be significant though, so I think you should still include it.

```
Present            Input Click Present Input         Present              Displayed
|--MsInPresentApi--|
|--MsBetweenPresents-----------|--MsBetweenPresents--|--MsUntilDisplayed--|

Click-to-Display:  |------------------------------------------------------|
= MsBetweenPresents + MsUntilDisplayed + prev(MsBetweenPresents) - prev(prev(MsInPresentApi))
```

**Comment by RodroG-85** [comment]
> Therefore, from a practical point of view, in the context of my analysis, and without modifying the equation for the calculation of the approximate input lag, should I only consider the results relative to the NVCP limiter as valid or useful?

@JeffersonMontgomery-Intel Could help me to solve this? Thanks in advance.

**Comment by aggieNick02** [comment]
Ok, that is good to know @JeffersonMontgomery-Intel. So even if you're looking at button click latency, if you want the upper bound to be as tight as possible, MsInPresentApi still comes in to play since it can be significant.

@RodroG-85 - I'm not sure how "solvable" it is. There are a couple different kinds of latency (continuous vs click input) that one can really look at here. They both are meaningful in different ways. If it is click input you are looking at, I think the best you can do is the upper bound we settled on:
MsBetweenPresents + MsUntilDisplayed + prev(MsBetweenPresents) - prev(prev(MsInPresentApi))

This upper bound will be correct regardless of which rate limiting method you use.

If you are looking at the more "continuous" latency - e.g. VR headset position lag - then the analysis above by @jenatali holds:

> In summary:
> 
> * The NVCP latency should be (more or less) accurate according to that equation.
> * RTSS is definitely not accurate according to that equation.
> * For in-engine limiters, well, each engine can implement it themselves, so it depends whether they sleep:
>   
>   * After rendering and before presenting: your equation would be accurate, and show a high latency.
>   * After presenting before rendering the next frame: your equation would be _innacurate_, and show a high latency.

**Comment by RodroG-85** [comment]
Thanks @aggieNick02 for your reply. I'm looking to solve this:

> A user on reddit compared different frame limiters, including NV CP and RTSS. CapFrameX provides input lag approximation based on your formular given in the GitHub description.
> LatencyMs =~ MsBetweenPresents + MsUntilDisplayed - previous(MsInPresentAPI)
> The problem is that the values for MsInPresentAPI are very high, so the overall latency becomes very low. This is only the case if the frame limiter from Nvidia is used. What's going wrong here? Is there a problem at all?
> Capture data with NV CP frame limiter: https://pastebin.com/0ipbcaKs
> Capture data with RTSS frame limiter: https://pastebin.com/ueSpNSWt

For me, it's clear that what @jenatali replied would the best answer and approach for my benchmarking purposes.

However, if we don't know how each in-engine limiter implements its own fps limiter hook, I understand that the calculations here wouldn't be valid in relation to the original approximate input lag equation either.

Therfore, in the end, I'd have a case, RTSS's, with clearly innacurate latency results; another, In-Engine, that is doubtful (and not useful for my benchmarking purposes), and only one, NV CP limiter, that would be correct (an useful for my analysis) in relation to the results gathered by applying the original and unmodified approximate input lag equation.

So, is the above tentative practical approach correct for [my analysis](https://www.reddit.com/r/allbenchmarks/comments/f5l6pk/nvidias_control_panel_fps_limiter_vs_rivatuner_vs/) and benchmarking purposes?

**Comment by JeffersonMontgomery-Intel** [comment]
Yes.

**Comment by RodroG-85** [comment]
Thank you.

**Comment by DevTechProfile** [comment]
I will change the approach in CapFrameX and consider the click input latency instead. So what I would suggest is to calculate the average of the lower and upper bound:

click input lantency ~= (MsUntilDisplay + MsBetweenPresents + MsBetweenPresents + MsUntilDisplayed + prev(MsBetweenPresents) - prev(prev(MsInPresentApi)) ) / 2 = (2 * MsBetweenPresents + 2 * MsUntilDisplayed + prev(MsBetweenPresents) - prev(prev(MsInPresentApi)) ) / 2  = MsBetweenPresents + MsUntilDisplayed + 0.5 * prev(MsBetweenPresents) - 0.5 * prev(prev(MsInPresentApi)) 

What do you think?

**Comment by JeffersonMontgomery-Intel** [comment]
I think you're missing one term:

`MsBetweenPresents + MsUntilDisplayed + 0.5 * prev(MsBetweenPresents) - 0.5 * prev(MsInPresentAPI) - 0.5 * prev(prev(MsInPresentApi))`

But yeah that should be the expected value of the click latency if a click is equally likely to happen at any time (and we assume game sampling input at beginning of the frame and ignoring HW latency etc.).

**Comment by DevTechProfile** [comment]
> I think you're missing one term:

lower bound = MsUntilDisplay + MsBetweenPresents
upper bound =  MsBetweenPresents + MsUntilDisplayed + prev(MsBetweenPresents) - prev(prev(MsInPresentApi))

That adds up to what I calculated above?! Do you assume 
MsUntilDisplay + MsBetweenPresents + prev (MsBetweenPresents)
as the valid lower bound ?

**Comment by JeffersonMontgomery-Intel** [comment]
Yes, the original equation is the lower bound.

**Comment by DevTechProfile** [comment]
After careful consideration, I do not think it makes sense to include the lower bound in the calculation. After all, in many cases this is incorrect. To combine something wrong with something right leads to something wrong. So I will only use the upper bound.

**Comment by JeffersonMontgomery-Intel** [comment]
There are certainly reasons to care about the maximum possible latency, so if that's what you need for your analysis, then great.

Another option (probably the most informative) is to show both the lower and upper bound.

In terms of expected latency (i.e., in the probabilistic sense as described here https://en.wikipedia.org/wiki/Expected_value) you want the average of the two because a click may occur anywhere between the lower and upper bounds, with equal probability.

**Comment by xcasxcursex** [comment]
TLDR Help us Intel_wan_kenobis, you're our only hope. 
Short version in bold to save you some time.


Gents, this issue and the new visualisation tools reaching a wider audience, have raised questions amongst the community, and while I have my theories, I know when I'm in the company of people who know better than I do. I hope you might spare us a moment of your thoughts, to help clear up some confusion.

Tests with high speed cameras have all shown, that driver-based and external framerate limiters, provide input latency which appears to be a full frame time longer, than in-game limiters (This is a generalisation, but one that applies to most cases - yes, there have been a few in-game limiters with higher latency in the region of external limiters, but I'm not aware of any external limiters which can match the latency of the 'faster' in-game limiters.) 

**Using the approximation metric here, in-game limiters, and driver or external limiters (eg RTSS), show essentially similar results. Being that this is in contradiction to high-speed camera observations, for some people this has raised doubts as to the validity of the approximation.** 

I think it is obvious that there are unknown latencies outside of the software realm (switch debounce times, USB polling, monitor redraw times, etc etc) and there are also variations in software (such as CPU queue depth), however on a given test machine and settings, these tend to be fairly consistent and just become a fairly static 'overhead'  - which is why the CapFrameX tool allows the user to simply add a few milliseconds onto the approximation.

Of course, this fairly static additional time does not explain the 'one extra frame' of latency we see in external/driver-based limiters. **Do you have any idea what might be causing that extra time for any limiter which is not in-game?**


Apologies for going slightly offtopic here, but I hope that by sharing some of your knowledge with the community, you might avert what is sure to be many internet arguments.....

**Comment by jenatali** [comment]
I can't think of a technical explanation without seeing data. If you'd like to provide evidence that can be analyzed, the best form would be GPUView traces, which can be captured following some excellent [instructions from NVIDIA](https://nvidia.custhelp.com/app/answers/detail/a_id/3507/~/how-to-generating-an-event-trace-log-for-gpuview-to-troubleshoot-performance). The resulting traces zip very well, and can probably be directly uploaded here.

If you're using a recent version of GPUView, you should be able to use "log.cmd present" instead of just "log.cmd" to start it - this results in smaller files and less recording overhead by limiting the traces to just the ETW providers that PresentMon is interested in.

**Comment by xcasxcursex** [comment]
Thank you, I will upload examples of RTSS and ingame limiters shortly.

On a related note, MS have made log.cmd redundant recently, but the Windows Performance Recorder Profile intended to replace it was accidentally excluded from the distribution of WPR. They were kind enough to share it elsewhere, so here it is:
[log_wprp.zip](https://github.com/GameTechDev/PresentMon/files/4284198/log_wprp.zip)

**Comment by xcasxcursex** [comment]
Unfortunately the traces are just a little too large for github's taste; there's a 10MB limit, and they're about 15 when compressed. I only recorded for a few seconds, with the present profile, so I don't think I could make them much smaller I'm afraid. Happy to deliver them to you in any way which is convenient for you.

**Comment by jenatali** [comment]
Regarding log.wprp... that was me that shared it. You should be able to upload them to DropBox, OneDrive, etc, and share a link here.

**Comment by xcasxcursex** [comment]
Haha well there you go.... it's a good thing I said about how kind you are, talking about you behind your back to your face XD

Since I don't have any accounts with such filesharing services, I had to get a little creative.... Hope this works.

https://github.com/xcasxcursex/Tmp/raw/master/Ingame%20Limiter.7z
https://github.com/xcasxcursex/Tmp/raw/master/RTSS%20Limiter.7z

**Comment by aggieNick02** [comment]
I've got a high-level theory on why the external limiters are generally worse than in-game. For sake of argument let's assume the input sampling and Present calls happen in a single thread, with fan-out and collapse-in for rendering like @jenatali described.

If you're writing an in-game limiter, you know you are going to put delays between your calls to present. But you control and expect these delays, so while they are happening, you can still be dealing with the commands being input and using them to prepare the next frame. You can get work done while your present thread (mostly) idles, because you are the one doing the idling.

If you've got an external limiter running, the game is suddenly dealing with Present calls that take longer than usual. Instead of being able to utilize the idle time to do work on the frame one ahead of the next to be presented, the time is wasted because it is not under the developer's control, and occurs during Present. The input happening during the sleep isn't processed until the Present is done.

This explanation also jives with the fact that the metrics seen by PresentMon will look similar for both in-game and external. The time between presents, time until display, etc, are all similar. PresentMon can't know about the input sampling, and it is the hooked-in sleep that is introducing unexpected lag in the input sampling.

I'll admit I don't know a lot about game dev/engines, but it seems like a straightforward explanation. Feel free to poke holes in it. :-)

**Comment by jenatali** [comment]
Hm, from those traces I'm not really seeing much of a difference. In both cases, the CPU and GPU go idle for a short period each frame, meaning all work being done is going towards the *current* frame as opposed to putting work towards a *queued future* frame. The amount of time from submitting the first rendering work to the frame appearing on-screen are very similar.

The only theory I can propose is the same as @aggieNick02, the only difference between the two is when the input is sampled, where the app changes its behavior when it's limiting its own framerate vs when it's externally limited.

**Comment by xcasxcursex** [comment]
Thanks guys, really appreciate your taking the time to look into this.

Your theory matches brief commentary made by game devs in the past (that the game's knowledge that it is delaying, allows it to sample input appropriately, during the wait time), as well as with high-speed camera footage in scenarios such as using newer low-latency driver settings (AMD anti-lag/Nvidia ULL, where the driver delays the graphics processing pipeline until just-in-time, but still results in slightly higher latencies than using an in-game limiter to ensure future frames are not queued (as in my examples above))

Input latency, being a large contributor to performance in competitive gaming, becomes one of those subjects which can result in a lot of misunderstanding, so voices of authority such as yours really help to 'keep it real'. You've done the community a favour. My apologies again for side-tracking you, with thanks from us all.

**Comment by DevTechProfile** [comment]
The new approach (expected input lag, which is calculated from the upper and lower bound) can be tested with latest CapFrameX version: https://github.com/DevTechProfile/CapFrameX/releases

Just check out the "Synchronization" page in CX.

@JeffersonMontgomery-Intel Issue can be closed.

**Comment by DevTechProfile** [comment]
Thanks for this very interesting discussion.! Closed...

**Comment by Kein** [comment]
> The difference is based on where the delay is added.
> 
> * If a delay is added by the OS/driver while Present() is running, then the delay will be included in the MsInPresentAPI metric
> * If Present() is hooked, or the application does this itself, and the delay added before Present() is called then it will be included in the MsBetweenPresents metric.
> 
> In both of these cases, the latency equation above will be correct (assuming the workload samples input and simulation time when Present() returns).
> 
> If the delay is added after Present() returns, but before the application samples input/time, then the equation will be wrong (too large) since that assumption is then incorrect.

Is this still correct? I'm testing RTSS new hybrid scanline sync and my delay is like 0.3ms at 80FPS. According to the docs, the mode I chose is hooking before Present ("front edge sync") so the delay should be included properly?
![image](https://user-images.githubusercontent.com/623756/97433950-5713ba00-192f-11eb-9045-e339c4849506.png)

**Comment by JeffersonMontgomery-Intel** [comment]
Yes the quote is still correct; "Front edge sync" corresponds to the second bullet point.

**Comment by Kein** [comment]
Okay, so if Present latency should be correctly calculated by PresentMon why I'm getting the impossible 0.3ms?

**Comment by JeffersonMontgomery-Intel** [comment]
Hmm, I'm not sure.  Would it be possible to share an ETL trace that demonstrates this?  If so I can take a look...

> If you'd like to provide evidence that can be analyzed, the best form would be GPUView traces, which can be captured following some excellent [instructions from NVIDIA](https://nvidia.custhelp.com/app/answers/detail/a_id/3507/~/how-to-generating-an-event-trace-log-for-gpuview-to-troubleshoot-performance). The resulting traces zip very well, and can probably be directly uploaded here.
> 
> If you're using a recent version of GPUView, you should be able to use "log.cmd present" instead of just "log.cmd" to start it - this results in smaller files and less recording overhead by limiting the traces to just the ETW providers that PresentMon is interested in.

**Comment by Kein** [comment]
Yeah, I tried it earlier today beforehand but it wont run, even though I've downloaded the Win7SDK one and Nvidia guide is Win10 only. If there is any other way to get that trace let me know.
